rust/lance/src/io/exec/scalar_index_join.rs

```diff
@@ -0,0 +1,458 @@
+            Partitioning::UnknownPartitioning(1),
+            EmissionType::Incremental,
+            Boundedness::Bounded,
+        );
+
+        Ok(Self {
+            input,
+            dataset,
+            join_column,
+            index_name,
+            join_type,
+            scalar_index: OnceLock::new(),
+            output_schema,
+            properties,
+        })
+    }
+
+    /// Load the scalar index (cached after first call)
+    async fn get_or_load_index(&self) -> Result<&Arc<dyn ScalarIndex>> {
+        if let Some(index) = self.scalar_index.get() {
+            return Ok(index);
+        }
+
+        // First load the index metadata by name to get the UUID
+        let idx = self
+            .dataset
+            .load_scalar_index(ScalarIndexCriteria::default().with_name(&self.index_name))
+            .await?
+            .ok_or_else(|| Error::Index {
+                message: format!("Index with name {} does not exist", self.index_name),
+                location: location!(),
+            })?;
+
+        // Then open the index using the UUID
+        let index = self
+            .dataset
+            .open_scalar_index(
+                &self.join_column,
+                &idx.uuid.to_string(),
+                &NoOpMetricsCollector,
+            )
+            .await?;
+
+        // Cache it
+        let _ = self.scalar_index.set(index);
+        // Return the cached index (either the one we just set or one set by another thread)
+        Ok(self.scalar_index.get().unwrap())
+    }
+
+    /// Perform index lookups for a batch of keys
+    #[instrument(skip_all, fields(num_keys = keys.len()))]
+    async fn lookup_keys(&self, keys: &dyn Array) -> Result<Vec<Option<u64>>> {
+        let index = self.get_or_load_index().await?;
+        let mut results = Vec::with_capacity(keys.len());
+
+        for i in 0..keys.len() {
+            if keys.is_null(i) {
+                // Null keys never match
+                results.push(None);
+                continue;
+            }
+
+            // Convert array value to ScalarValue
+            let scalar_value = ScalarValue::try_from_array(keys, i)?;
+
+            // Query the index
+            let query = SargableQuery::Equals(scalar_value);
+            let search_result = index.search(&query, &NoOpMetricsCollector).await?;
+
+            // Extract the first matching row ID (if any)
+            let row_id = match search_result {
+                SearchResult::Exact(row_ids) => {
+                    if row_ids.is_empty() {
+                        None
+                    } else {
+                        // For merge insert, we expect at most one match per key
+                        // If there are multiple, take the first one
+                        row_ids
+                            .row_ids()
+                            .and_then(|mut iter| iter.next())
+                            .map(u64::from)
```

Comment:
I think we've discussed before but this seems like a horrible situation to me. I almost wonder if this should be an error. If this happens then the user's data is almost certainly corrupt and the merge insert will worsen the corruption.

Reply:
That's reasonable. It does error by default in Databricks [^1] and Snowflake [^2].

So we can start to enforce that in this PR. We can enforce that in other code paths as a follow up:
* https://github.com/lancedb/lance/issues/4535

[^1]: https://docs.databricks.com/aws/en/sql/language-manual/delta-merge-into#:~:text=MERGE%20operations%20fail%20with%20a%20DELTA_MULTIPLE_SOURCE_ROW_MATCHING_TARGET_ROW_IN_MERGE
[^2]: https://docs.snowflake.com/en/sql-reference/sql/merge#nondeterministic-results-for-update-and-delete